### PR TITLE
feat: auto-attach $screen_name to all events

### LINKED ISSUE #119: Set `$screen_name` to all events
### Description

- [ ] Android
- [ ] iOS
- [X] Flutter (when calling `screen` via PosthogObserver or manually)
- [X] RN (when calling `screen` via wix, react-navigation or manually)

### PR DESCRIPTION
**Pull request by turnipdabeets**
## :bulb: Motivation and Context

Closes PostHog/posthog-android#119 for iOS. RN and Flutter already attach `$screen_name` to every event once `screen()` has been called; iOS did not. Android equivalent: PostHog/posthog-android#530.

## :green_heart: How did you test it?

Six unit tests in `PostHogScreenNameTest`: before-screen baseline, after-screen attaches, caller override wins, `reset()` clears, `$exception` carries it (Option A cross-event scope), `$snapshot` does not (verifies `appendSharedProps = false` gate).

End-to-end on iPhone 17 simulator across all five customer-visible scenarios — confirmed via Xcode console and by querying the `events` table on `us.posthog.com`:

| Scenario | Event | `$screen_name` |
|---|---|---|
| Auto-capture (`captureScreenViews=true`) | `test_auto_capture` | `UIHostingController<ModifiedContent<AnyView, RootModifier>>` |
| Manual `screen("ManualScreen")` | `test_after_manual` | `ManualScreen` |
| Caller override | `test_override` | `OverrideValue` |
| After `reset()` | `test_after_reset` | _absent_ |
| Exception after `screen("ExceptionScreen")` | `$exception` | `ExceptionScreen` |

**Behavior change** for reviewers: existing customers who have not disabled `captureScreenViews` will start seeing `$screen_name` attached to `$exception`, `$identify`, etc. — not only `$screen` events. Opt-out is `captureScreenViews = false` plus avoiding manual `PostHogSDK.shared.screen()` calls.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [ ] No breaking change or entry added to the changelog.

## If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

### CODE REVIEW

File: PostHog/PostHogSDK.swift [hunk 1]
```diff
@@ -433,6 +439,13 @@ let maxRetryDelay = 30.0
             }
 
             props["$process_person_profile"] = hasPersonProcessing()
+
+            // Only stamp if the caller didn't supply one — `merging(properties)`
+            // below keeps the existing value on conflict, so seeding would
+            // shadow a caller-supplied override.
+            if let name = lastScreenName, properties?["$screen_name"] == nil {
```
Comment (ioannisj) [inline PostHog/PostHogSDK.swift]: We should be testing for .isEmpty as well, otherwise we may pass an empty string in the props?

File: PostHog/PostHogSDK.swift [hunk 2]
```diff
@@ -1958,6 +1986,7 @@ let maxRetryDelay = 30.0
             didEnterBackgroundToken = nil
             logger?.detach()
             logger = nil
+            lastScreenLock.withLock { _lastScreenName = nil }
```
Comment (ioannisj) [inline PostHog/PostHogSDK.swift]: I think we better move this outside setupLock? Just for symmetry today, but if in the future these locks are acquired in the reverse order somewhere else this may deadlock

File: PostHogTests/PostHogScreenNameTest.swift [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+            sut.reset()
+            sut.close()
+        }
+
+        it("caller-supplied screen_name overrides cached value") {
+            let sut = self.getSut(captured: captured)
+
+            sut.screen("Home")
+            sut.capture("event", properties: ["$screen_name": "Override"])
```
Comment (ioannisj) [inline PostHogTests/PostHogScreenNameTest.swift:72]: Should we also add a test for screen() override? 
```
screen("Home", properties: ["$screen_name": "Override"]) 
```
This is expected to carry `"$screen_name": "Override"` right?

File: PostHog/PostHogSDK.swift [hunk 4]
```diff
@@ -38,6 +38,12 @@ let maxRetryDelay = 30.0
     private let cachedPersonPropertiesLock = NSLock()
     private var cachedPersonPropertiesHash: String?
 
+    private let lastScreenLock = NSLock()
+    private var _lastScreenName: String?
+    var lastScreenName: String? {
```
Comment (ioannisj) [inline PostHog/PostHogSDK.swift:43]: Can this be private? Do we need it to be internal?

Comment (ioannisj) [inline PostHog/PostHogSDK.swift:43]: PostHogLogger maintains its own `lastScreenName`. Could we reuse? Maybe a PostHogScreenNameCache holder so both logger and SDK instance can read from that, or update Logger to read from here?

Comment (ioannisj) [reply PostHog/PostHogSDK.swift:43]: Depends on https://github.com/PostHog/posthog-ios/pull/609#discussion_r3290700491 I guess

File: PostHog/PostHogSDK.swift (hunk 1 above)
Comment (turnipdabeets) [reply PostHog/PostHogSDK.swift]: Done in 3bd547f4c — added `!name.isEmpty` on the cached value and `isNilOrEmpty` on the caller-supplied value (treats whitespace-empty as absent so we still fall back to the cache).

File: PostHog/PostHogSDK.swift (hunk 2 above)
Comment (turnipdabeets) [reply PostHog/PostHogSDK.swift]: Done in 3bd547f4c — moved the `lastScreenLock` clear outside the `setupLock` block, with a one-line comment explaining the lock-ordering motivation.

File: PostHog/PostHogSDK.swift (hunk 4 above)
Comment (turnipdabeets) [reply PostHog/PostHogSDK.swift:43]: Done in 3bd547f4c — tightened to `private`. Nothing outside this file reads it.

File: PostHogTests/PostHogScreenNameTest.swift (hunk 3 above)
Comment (turnipdabeets) [reply PostHogTests/PostHogScreenNameTest.swift:72]: Added in 4fab84cdd. Also aligned iOS to match Android — caller-supplied `$screen_name` in `properties` now wins on the `$screen` event (previously the `screenTitle` arg won via the iOS-specific `{ prop, _ in prop }` closure). The new test asserts `$screen_name == "Override"` after `screen("Home", properties: ["$screen_name": "Override"])`.

File: PostHog/PostHogSDK.swift (hunk 4 above)
Comment (turnipdabeets) [reply PostHog/PostHogSDK.swift:43]: Done in 4fab84cdd. SDK now owns the single cache; `PostHogLogger.lastScreenName` delegates to `PostHogSDK.lastScreenName`. `screen()` sanitizes once and the (sanitized) value goes to both the `$screen` event payload and the cache. Logger no longer subscribes to the screen-view publisher and `detach()` was removed (dead code). Side effect of doing the consolidation this way: SwiftUI `$screen` events now carry the sanitized name too — flagged in the changeset.

Comment (ioannisj) [reply PostHog/PostHogSDK.swift:43]: > SwiftUI $screen events now carry the sanitized name too — flagged in the changeset.

This will probably affect analytics dashboards/queries for people but I think it's a sensible change

**Review by ioannisj — APPROVED** [review]

File: PostHog/PostHogSDK.swift (hunk 4 above)
Comment (turnipdabeets) [reply PostHog/PostHogSDK.swift:43]: True, but the string before probably wasn't very helpful. The solution is to use `.postHogScreenView("MyView")`

File: PostHog/PostHogSDK.swift [hunk 5]
```diff
@@ -1148,13 +1177,30 @@ let maxRetryDelay = 30.0
             return
         }
 
+        // Strip SwiftUI wrappers (UIHostingController<X> → X). Drops degenerate
+        // inputs (empty, stripped-AnyView) — for those we skip the $screen event
+        // and leave the cache untouched so the last useful name survives.
+        guard let cleaned = PostHogLogger.sanitize(rawScreenName: screenTitle) else {
```
Comment (marandaneto) [inline PostHog/PostHogSDK.swift]: strange to have the attribute declared in this class, and the sanitize in the PostHogLogger class

File: PostHog/Logs/PostHogLogger.swift [hunk 6]
```diff
@@ -9,51 +9,38 @@ import Foundation
...
     private weak var sdk: PostHogSDK?
-    private let lastScreenLock = NSLock()
-    private var _lastScreenName: String?
-    private var screenViewToken: RegistrationToken?
 
-    /// Latest reported screen name, populated by the screen-view publisher.
-    /// `nil` until the first navigation after SDK setup.
+    /// Latest reported screen name. Delegates to `PostHogSDK.lastScreenName`,
+    /// which is sanitized in `screen()` before being cached.
     var lastScreenName: String? {
```
Comment (marandaneto) [inline PostHog/Logs/PostHogLogger.swift]: probably lastScreenName should live in a single place and shared with this class and the posthogsdk instead, right now it feels its dupe

File: PostHog/PostHogSDK.swift (hunk 5 above)
Comment (turnipdabeets) [reply PostHog/PostHogSDK.swift]: updated

File: .changeset/screen-name-on-all-events.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+- **`screen("")` is silently dropped.** Previously an empty `screenTitle` emitted a `$screen` event with `$screen_name = ""`; now nothing is emitted and the cache is untouched. Customers using empty-string as a sentinel in dashboards will see those rows disappear.
+- **`screen("AnyView")` (literal manual call) is honored.** Only `AnyView` that surfaced from stripping `UIHostingController<...>` / `ModifiedContent<...>` wrappers is dropped (auto-capture noise from `body: some View` type erasure). A caller who literally typed `screen("AnyView")` gets that event through.
+- **Out-of-process crash reports carry `$screen_name` too.** `screen()` now refreshes the crash-replay context snapshot, so an `$exception` event reconstructed from a crash that happened on a given screen will ship with that screen name — matching the behavior of live `captureException(_:)` calls.
```
Comment (marandaneto) [inline .changeset/screen-name-on-all-events.md]: https://github.com/PostHog/posthog-android/pull/530/changes#r3304363607

**Review by marandaneto — APPROVED** [review]
